samza-core/src/main/java/org/apache/samza/clustermanager/StandbyContainerManager.java

```diff
@@ -375,24 +421,39 @@ boolean checkStandbyConstraints(String containerIdToStart, String host) {
       SamzaResource resource = samzaApplicationState.pendingProcessors.get(containerID);
 
       // return false if a conflicting container is pending for launch on the host
-      if (resource != null && resource.getHost().equals(host)) {
-        log.info("Container {} cannot be started on host {} because container {} is already scheduled on this host",
-            containerIdToStart, host, containerID);
+      if (!checkStandbyConstraintsHelper(containerIdToStart, host, resource, containerID)) {
         return false;
       }
 
       // return false if a conflicting container is running on the host
       resource = samzaApplicationState.runningProcessors.get(containerID);
-      if (resource != null && resource.getHost().equals(host)) {
-        log.info("Container {} cannot be started on host {} because container {} is already running on this host",
-            containerIdToStart, host, containerID);
+      if (!checkStandbyConstraintsHelper(containerIdToStart, host, resource, containerID)) {
         return false;
       }
     }
 
     return true;
   }
 
+  boolean checkStandbyConstraintsHelper(String containerIdToStart, String hostToStartContainerOn, SamzaResource existingResource, String existingContainerID) {
+    if (existingResource != null) {
+      ClusterManagerConfig clusterManagerConfig = new ClusterManagerConfig(config);
+      if (clusterManagerConfig.getFaultDomainAwareStandbyEnabled() && faultDomainManager.hasSameFaultDomains(hostToStartContainerOn, existingResource.getHost())) {
+        log.info("Container {} cannot be started on host {} because container {} is already scheduled on this rack",
+                containerIdToStart, hostToStartContainerOn, existingContainerID);
+        if (StandbyTaskUtil.isStandbyContainer(containerIdToStart)) {
+          samzaApplicationState.failedFaultDomainAwareContainerAllocations.incrementAndGet();
+        }
+        return false;
+      } else if (existingResource.getHost().equals(hostToStartContainerOn)) {
+        log.info("Container {} cannot be started on host {} because container {} is already scheduled on this host",
+                containerIdToStart, hostToStartContainerOn, existingContainerID);
```

Comment:
seems like the logging is now impacted and different from the previous flow in the absence of rack aware standby.
i.e. this method gets invoked with `pendingResource` and `runningResource` and seems both will end up printing as `scheduled on this host`

Reply:
that is correct. the logging message is changed and it also means semantic difference during debugging to differentiate  if its pending code path that triggered this vs running code path that triggered this

Reply:
Thanks for noticing the logging difference. I've added a parameter to the helper method to log differently in each case.